Deploy Engine and Replica Manager Pods

ttpcodes opened the pull request:
This is part of the ongoing work to refactor the `longhorn-engine`.

This PR modifies the logic that the `Node Controller` uses to reconcile its state to additionally deploy its own `Pods` that are needed for managing all the `Engine` and `Replica` instances on the `Node`. As part of the reconciliation logic, the `Node Controller` additionally checks the known `Engine Images` to see if there are `Engine` and/or `Replica` `Manager Pods`deployed. If either of these aren't deployed, the controller will take care of deploying these `Pods` on its own `Node`. 

The `Pods` have `Labels` attached to it denoting its type and associated `Engine Image` so that the `Controller` has a way to refer to it when checking to see if the `Pods` have been deployed. Additionally, the `Pods` have `Owner References` to the `Engine Image` it is associated with, and will be cleaned up by `Kubernetes'` garbage collector once the associated `Engine Image` is deleted.

The `Node Controller` also has an additional `Engine Image Informer` to keep track of changes to the `Engine Images`.

Note that as part of the refactor, this change will result in old versions of any `Engine Images` that do not support the new `Launcher` architecture to deploy unsuccessfully in these new `Pods`, entering a `Crash Loop Back Off`.

ttpcodes commented on controller/node_controller.go:
```diff
@@ -731,3 +765,169 @@ func (nc *NodeController) syncNodeStatus(pod *v1.Pod, node *longhorn.Node) error
+		if err != nil {
+			return errors.Wrapf(err, "cannot get replica launcher pod for engine image %v", id)
+		}
+		if replicaPod == nil && len(node.Spec.Disks) > 0 {
```
We're intending on deploying the `Replica Manager Pod` only if the `Node` has `Disks`, since otherwise there's nowhere to store the `Replicas`.

However, what happens if the `Disks` get removed after this `Pod` has already been created? How should we address this?

yasker replied:
Yes, we need to address this. The replica pod lifecycle is bound with the availability of the disk. One disk can only be removed if there is no replica stored in it, so it should be safe to remove the replica pod as well.

ttpcodes replied:
Adding logic to delete the `Replica Pod` once no `Disks` remain on the `Node`.

ttpcodes commented:
- Added logic to automatically delete the `Replica Manager Pod` once a `Node` reports that no `Disks` remain on it.
- Rebased onto `master`.
- Converted `CreateNodeEnginePod` to a generic `CreatePod` function that can be utilized in other cases.

ttpcodes commented:
- Created a new function called `NodeController.createGenericManagerPodSpec` to avoid repeating a majority of the `Pod` spec for the `Engine Pod` and `Replica Pod`.
- Added a readiness check in response to the new health check server in `longhorn-engine-launcher`.

ttpcodes commented:
Closing this PR until all the new changes to the refactor design have been implemented. Changes can be tracked on my fork of `longhorn-manager` under the branch `engine-launcher-refactor-dev`.
